plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+
+        dt_range = [0, np.inf] * u.s if dt_range is None else dt_range
+        self.dt_range = dt_range.to(u.s).value
+
+        # Update the `tracker` attribute so that the stop condition & save routine can be used
+        termination_condition.tracker = self
+
+        if save_routine is not None:
+            save_routine.tracker = self
+
+        self.termination_condition = termination_condition
+        self.save_routine = save_routine
+
+    @property
+    def num_grids(self):
+        """The number of grids specified at instantiation."""
+        return len(self.grids)
+
+    def setup_adaptive_time_step(
+        self,
+        time_steps_per_gyroperiod: Optional[int] = 12,
+        Courant_parameter: Optional[float] = 0.5,
+    ):
+        """Set parameters for the adaptive time step candidates.
+
+        Parameters
+        ----------
+        time_steps_per_gyroperiod : int, optional
+            The number of subdivisions of the particle's gyration. The default is twelve.
+
+        Courant_parameter : float, optional
+            The Courant parameter is the fraction of a grid cell length by which to subdivide
+            the motion of the particle.
+
+
+        Notes
+        -----
+        Two candidates are calculated for the adaptive time step: a time step based on the gyroradius
+        of the particle and a time step based on the resolution of the grid. The candidate associated
+        with the gyroradius of the particle takes a ``time_steps_per_gyroperiod`` parameter that specifies
+        how many times the orbit of a gyrating particles will be subdivided. The other candidate,
+        associated with the spatial resolution of the grid object, calculates a time step using the time
+        it would take the fastest particle to cross some fraction of a grid cell length. This fraction is the Courant number.
+        """
+
+        if not self.is_adaptive_time_step:
+            raise ValueError(
+                "The setup adaptive time step method only applies to adaptive time steps!"
+            )
+
+        self._steps_per_gyroperiod = time_steps_per_gyroperiod
+        self._Courant_parameter = Courant_parameter
+
+    def _log(self, msg):
+        if self.verbose:
+            print(msg)  # noqa: T201
+
+    @particle_input
+    def load_particles(
+        self,
+        x,
+        v,
+        particle: Particle = Particle("p+"),  # noqa: B008
+    ):
+        r"""
+        Load arrays of particle positions and velocities.
+
+        Parameters
+        ----------
+        x : `~astropy.units.Quantity`, shape (N,3)
+            Positions for N particles
+
+        v : `~astropy.units.Quantity`, shape (N,3)
+            Velocities for N particles
+
+        particle : |particle-like|, optional
+            Representation of the particle species as either a |Particle| object
+            or a string representation. The default particle is protons.
+
+        distribution : str
```

Comment:
Delete 'distribution` parameter from docstring, that's from the proton radiography class

Reply:
I should have caught that. My bad